services/supervisor/src/lib.rs

```diff
@@ -150,12 +154,10 @@ impl Service for Supervisor {
                     }
                 };
                 // TODO Rewrite on async await syntax. [ECR-3222]
-                context.dispatcher_channel().request_deploy_artifact(
-                    context.instance.id,
-                    artifact,
-                    spec,
-                    and_then,
-                );
+                let action = Action::StartDeploy { artifact, spec };
```

Comment:
Off-topic: Do I understand correctly that each validator triggers a deploy action and broadcasts endorsements of all pending deploys on each block? Is this the expected behavior? Are the repeated endorsement transactions actually broadcast, or is this suppressed somewhere?

Reply:
Yes, you understand that (almost) correctly.
It is *somewhat* suppressed: deployment is currently applied synchronously right after request (it's in `after_commit` and request is processed after `after_commit` event), so every validator sends a confirmation in the next block, and thus it will be filtered in the `filter` above.

One piece of flawed code causes other piece of flawed code to work correctly. Checkmate, adepts of clean programming :)

Talking seriously, yeah, it's a breach. And the current logic of deploy-register-init process in Supervisor should be refined as well.